src/components/KTreeList/KTreeList.vue

```diff
@@ -0,0 +1,458 @@
+import draggable from 'vuedraggable'
+import KTreeItem, { TreeListItem } from '@/components/KTreeList/KTreeItem.vue'
+
+export interface ChangeEvent {
+  items: TreeListItem[]
+}
+
+export interface ChildChangeEvent {
+  parent: string,
+  children: TreeListItem[]
+}
+
+const props = defineProps({
+  modelValue: {
+    type: Array as PropType<TreeListItem[]>,
+    default: null,
+    validator: (items: TreeListItem[]) => !items.length || items.every(i => i.name !== undefined && i.id !== undefined),
+  },
+  items: {
+    type: Array as PropType<TreeListItem[]>,
+    default: null,
+    validator: (items: TreeListItem[]) => !items.length || items.every(i => i.name !== undefined && i.id !== undefined),
+  },
+  disableDrag: {
+    type: Boolean,
+    default: false,
+  },
+  maxLevel: {
+    type: Number,
+    default: 2,
+  },
+  /**
+   * These are internal props used to track info
+   * of nested children.
+   */
+  level: {
+    type: Number,
+    default: 0,
+  },
+  parentId: {
+    type: String,
+    default: undefined,
+  },
+})
+
+const emit = defineEmits<{
+  (event: 'change', data: ChangeEvent): void,
+  (event: 'child-change', data: ChildChangeEvent): void,
+  (event: 'start'): void,
+  (event: 'end'): void,
+  (event: 'selected', item: TreeListItem): void
+}>()
+
+const key = ref(0)
+const internalList = ref<TreeListItem[]>([])
+const draggableAttrs = {
+  tag: 'div',
+  // Note: force-fallback: true is required to override cursor when dragging
+  'force-fallback': true,
+  animation: '100',
+  draggable: '.k-tree-item-container',
+  'item-key': 'id',
+  'ghost-class': 'k-tree-item-dragged',
+  'drag-class': 'k-tree-item-grabbing',
+  class: 'child-drop-zone',
+}
+const dragging = ref(false)
+// using these vars to trick the TS compiler and avoid errors about circular refs in the <template>
+const itemIcon = 'item-icon'
+const itemLabel = 'item-label'
+
+// we need this so we can create a watcher for programmatic changes to the modelValue
+const value = computed({
+  get(): TreeListItem[] {
+    return props.modelValue
+  },
+  set(newValue: TreeListItem[]): void {
+    internalList.value = newValue
+  },
+})
+
+const hasChildren = (item: TreeListItem): boolean => {
+  return !!internalList.value.filter(anItem => anItem.id === item.id)?.[0].children?.length
+}
+
+const indentStyle = computed((): Record<string, any> => {
+  const level1offset = props.level === 1 ? 6 : 0
+  return {
+    marginLeft: 8 * props.level + level1offset + 'px',
+  }
+})
+
+const iconSecondaryColor = (item: TreeListItem): string | undefined => {
+  if (item.icon === 'treeDoc' || !item.icon) {
+    return item.selected ? 'var(--teal-200)' : 'var(--grey-200)'
+  }
+
+  return undefined
+}
+
+const handleSelection = (itemToSelect: TreeListItem, list?: TreeListItem[]): void => {
+  if (!list) { // root level
+    // select the item
+    const selectedItem = internalList.value.filter((item: TreeListItem) => item.id === itemToSelect.id)?.[0]
```

Comment:
Quick one: Isn't `TreeListItem` what we need here? Could we avoid the filtering in that case?